server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -160,4 +384,478 @@ private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request)
+        this.performPhaseOnShard(
+            sidx,
+            shardIterators[sidx],
+            new SearchShardTarget(targetNode.nodeId(), shard.shardId, targetNode.clusterAlias())
+        );
+    }
+
+    private void onNodeQueryFailure(Exception e, NodeQueryRequest request, CanMatchPreFilterSearchPhase.SendingTarget target) {
+        for (ShardToQuery shard : request.shards) {
+            int idx = shard.shardIndex;
+            onShardFailure(idx, new SearchShardTarget(target.nodeId(), shard.shardId, target.clusterAlias()), shardIterators[idx], e);
+        }
+    }
+
+    public static final String NODE_SEARCH_ACTION_NAME = "indices:data/read/search[query][n]";
+
+    private static final CircuitBreaker NOOP_CIRCUIT_BREAKER = new NoopCircuitBreaker("request");
+
+    static void registerNodeSearchAction(
+        SearchTransportService searchTransportService,
+        SearchService searchService,
+        SearchPhaseController searchPhaseController
+    ) {
+        var transportService = searchTransportService.transportService();
+        var threadPool = transportService.getThreadPool();
+        final Dependencies dependencies = new Dependencies(searchService, threadPool.executor(ThreadPool.Names.SEARCH));
+        final int searchPoolMax = threadPool.info(ThreadPool.Names.SEARCH).getMax();
+        transportService.registerRequestHandler(
+            NODE_SEARCH_ACTION_NAME,
+            EsExecutors.DIRECT_EXECUTOR_SERVICE,
+            NodeQueryRequest::new,
+            (request, channel, task) -> {
+                final CancellableTask cancellableTask = (CancellableTask) task;
+                final int shardCount = request.shards.size();
+                int workers = Math.min(request.searchRequest.getMaxConcurrentShardRequests(), Math.min(shardCount, searchPoolMax));
+                final var state = new QueryPerNodeState(
+                    new QueryPhaseResultConsumer(
+                        request.searchRequest,
+                        dependencies.executor,
+                        NOOP_CIRCUIT_BREAKER, // noop cb for now since we do not have a breaker in this situation in un-batched execution
+                        searchPhaseController,
+                        cancellableTask::isCancelled,
+                        SearchProgressListener.NOOP,
+                        shardCount,
+                        e -> logger.error("failed to merge on data node", e)
+                    ),
+                    request,
+                    cancellableTask,
+                    channel,
+                    dependencies
+                );
+                // TODO: log activating or otherwise limiting parallelism might be helpful here
+                for (int i = 0; i < workers; i++) {
+                    executeShardTasks(state);
+                }
+            }
+        );
+        TransportActionProxy.registerProxyAction(transportService, NODE_SEARCH_ACTION_NAME, true, NodeQueryResponse::new);
```

Comment:
can you remind me what the proxy action is for? Did you get test failures without it?

Reply:
Yes I did get failures. 
It's for when you cannot directly send a message from node A to remote node B and the message needs to be relayed by an intermediary node C (in the same cluster as B). That relaying is done by the proxy action.